src/rust/engine/graph/src/entry.rs

```diff
@@ -79,6 +85,24 @@ impl<N: Node> EntryResult<N> {
       EntryResult::Clean(..) => true,
       EntryResult::Uncacheable(_, session_id) => context.session_id() == session_id,
       EntryResult::Dirty(..) => false,
+      EntryResult::UncacheableDependencies(..) => false,
+    }
+  }
+
+  fn has_uncacheable_deps(&self) -> bool {
+    match self {
+      EntryResult::Uncacheable(_, _) | EntryResult::UncacheableDependencies(_) => true,
+      EntryResult::Clean(..) | EntryResult::Dirty(..) => false,
+    }
+  }
+
+  /// Returns true if this result should block for polling (because there is no work to do
+  /// currently to clean it).
+  fn poll_should_wait(&self, context: &N::Context) -> bool {
+    match self {
+      EntryResult::Uncacheable(_, session_id) => context.session_id() == session_id,
```

Comment:
If we are polling an uncacheable node and the result was for the same session we are currently in, we wait for the uncacheable node to finish?

Reply:
`poll` waits for something to _change_ (unlike `waiters`, which are about waiting for something to finish). If the uncacheable node matches our session/run-id, then it is already up to date.